ignite/metrics/rouge.py

```diff
@@ -0,0 +1,151 @@
+
+import torch
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics import Metric
+
+# These decorators helps with distributed settings
+from ignite.metrics.metric import reinit__is_reduced, sync_all_reduce
+
+
+class Rouge(Metric):
+    r"""Calculates the Rouge Score for two Sequence of Tokens.
+
+
+    .. math::
+        F_\1 = \left( 1 - \alpha \right) * \frac{ \text{precision} * \text{recall} }
+        { \left( \alpha * \text{precision} \right) + \text{recall} }
+
+    where :math:`\alpha` is a float between 0 and 1.
+
+    - ``update`` must receive output of the form ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y}``.
+    - `y_pred` must be a Sequence of Tokens.
+    - `y` must be in the form of List of Sequence of Tokens of Model Texts.
+
+    Args:
+        alpha: alpha value for calculating f1 score
+        n: Rouge score to be calculated using n-grams of tokens
+        output_transform (callable, optional): a callable that is used to transform the
+            :class:`~ignite.engine.engine.Engine`'s ``process_function``'s output into the
+            form expected by the metric. This can be useful if, for example, you have a multi-output model and
+            you want to compute the metric with respect to one of the outputs.
+
+
+    """
+
+    def __init__(self, alpha: float, n: int, device: Union[str, torch.device] = torch.device("cpu")) -> None:
+        self._rougetotal: torch.Tensor = torch.tensor(0, device=device)
+        self._num_examples: int = 0
+        self._check_parameters(alpha, n)
+        self.alpha: float = alpha
+        self.n: int = n
+        super(Rouge, self).__init__(device=device)
+
+    def _check_parameters(self, alpha: float, n: int) -> None:
+        if alpha < 0 or alpha > 1:
+            raise ValueError("Alpha has to be between 0 and 1 to calculate Rouge f1 score.")
+        if isinstance(n, str) and n != "l" and n != "L":
+            raise ValueError('Invalid String, Only Rouge-L supported.Please use "l" or "L"')
+        elif isinstance(n, int) and n < 1:
+            raise ValueError("Ignite needs atleast unigram to calculate Rouge")
+
+    def _ngramify(self, text: List[str], n: int) -> Dict:
+        ngram_dict: DefaultDict = defaultdict(int)
+        start = 0
+        end = n
+        ngram = ""
+        for i in range(start, end):
+            ngram += text[i]
+            ngram += " "
+        while end < len(text):
+            ngram_dict[ngram] += 1
+            ngram = ngram[len(text[start]) + 1 :]
+            ngram += text[end]
+            ngram += " "
+            start += 1
+            end += 1
+        ngram_dict[ngram] += 1
+        return ngram_dict
+
+    def _safe_divide(self, numerator: float, denominator: float) -> float:
+        if denominator > 0:
+            return numerator / denominator
+        else:
+            return 0.0
+
+    def _f1_score(self, matches: int, recall_total: int, precision_total: int, alpha: float = 1) -> float:
```

Comment:
if `alpha = 0.5` we get `0.5 f_1`
if `alpha = 1` we get `precision`
if `alpha = 0` we get `recall`
This is not the `f_\alpha` in the reference paper, right ?

I think the reference implementation is https://github.com/bdusell/rougescore and this is a variation of the original paper. Do we want this or original algorithm ? I prefer provide the original algorithm.

Reply:
Implementation is inline with  https://github.com/google-research/google-research/blob/dbdf11998a865ba719b651310675f530c3c5763a/rouge/rouge_scorer.py#L125

Reply:
Only difference is the preprocessing between the  https://github.com/bdusell/rougescore  and the google research implementation. And we are leaving the preprocessing of text to user that's why we are taking sequence of tokens as input and not the original text. also at alpha = 0.5 we get original f1-score, but sometimes we might want more weight towards precision or recall that is why alpha is provided.

Reply:
https://github.com/google-research/google-research/blob/dbdf11998a865ba719b651310675f530c3c5763a/rouge/scoring.py#L162

Your implementation of `f1` is different.

Reply:
what I am saying is alpha = 0.5 is that formula. but you can change alpha depending on which is more important precision or recall. If you want I can remove alpha, and just add simple 2* (formula) instead. I thought giving slight flexibility to the user would be nice, since it does not require much code and would be nicer for a user. Also, I have deliberately made alpha a keyword argument so that if a user does not add alpha we can by default use the 2*(formula) instead.

Reply:
IMHO we should provide a clean implementation of the reference paper first and use right term for scores that we use.